File: dotnet/src/Rerout/Models/ProjectStats.cs
using System.Collections.Generic;
using System.Text.Json.Serialization;

namespace Rerout.Models;

/// <summary>Aggregate analytics for a project across the requested window.</summary>
public sealed record ProjectStats
{
    /// <summary>Window size in days the totals span.</summary>
    [JsonPropertyName("days")]
    public required int Days { get; init; }

    /// <summary>Total clicks recorded in the window.</summary>
    [JsonPropertyName("total_clicks")]
    public required long TotalClicks { get; init; }

    /// <summary>Total QR scans (subset of <see cref="TotalClicks"/>) in the window.</summary>
    [JsonPropertyName("qr_scans")]
    public required long QrScans { get; init; }

    /// <summary>One point per day across the window. Gap-filled by the server.</summary>
    [JsonPropertyName("daily")]
    public IReadOnlyList<DailyClicksPoint> Daily { get; init; } = [];

    /// <summary>Top countries by click count.</summary>
    [JsonPropertyName("countries")]
    public IReadOnlyList<StatsBreakdown> Countries { get; init; } = [];

    /// <summary>Top referrers by click count.</summary>
    [JsonPropertyName("referrers")]
    public IReadOnlyList<StatsBreakdown> Referrers { get; init; } = [];

    /// <summary>Click share by device class.</summary>
    [JsonPropertyName("devices")]
    public IReadOnlyList<StatsBreakdown> Devices { get; init; } = [];

    /// <summary>Click share by browser or in-app web view.</summary>
    [JsonPropertyName("browsers")]
    public IReadOnlyList<StatsBreakdown> Browsers { get; init; } = [];

    /// <summary>Top short codes by click count.</summary>
    [JsonPropertyName("top_codes")]
    public IReadOnlyList<StatsBreakdown> TopCodes { get; init; } = [];
}

/// <summary>Information about the project that owns the current API key.</summary>
public sealed record ProjectInfo
{
    /// <summary>Project identifier.</summary>
    [JsonPropertyName("id")]
    public required string Id { get; init; }

    /// <summary>Human-readable project name.</summary>
    [JsonPropertyName("name")]
    public required string Name { get; init; }

    /// <summary>URL-safe project slug.</summary>
    [JsonPropertyName("slug")]
    public required string Slug { get; init; }
}

/// <summary>A page of links returned by <see cref="Resources.Links.ListAsync"/>.</summary>
public sealed record ListLinksResult
{
    /// <summary>The links on this page.</summary>
    [JsonPropertyName("links")]
    public IReadOnlyList<Link> Links { get; init; } = [];

    /// <summary>
    /// Pass to the next <see cref="Resources.Links.ListAsync"/> call to fetch the
    /// next page. <c>null</c> when there are no more pages.
    /// </summary>
    [JsonPropertyName("next_cursor")]
    public long? NextCursor { get; init; }
}

/// <summary>Result of a delete call.</summary>
public sealed record DeleteResult
{
    /// <summary><c>true</c> when the link was deleted.</summary>
    [JsonPropertyName("deleted")]
    public required bool Deleted { get; init; }
}

File: java/src/main/java/co/rerout/sdk/model/ProjectStats.java
/*
 * rerout-java — Official Java SDK for the Rerout branded-link API.
 *
 * Copyright (c) 2026 Codecraft Solutions
 * Licensed under the MIT License — https://codecraftsolutions.co.za
 */

package co.rerout.sdk.model;

import com.google.gson.annotations.SerializedName;
import java.util.Collections;
import java.util.List;

/** Aggregate analytics for a project across the requested window. */
public final class ProjectStats {

    private final int days;

    @SerializedName("total_clicks")
    private final long totalClicks;

    @SerializedName("qr_scans")
    private final long qrScans;

    private final List<DailyClicksPoint> daily;
    private final List<StatsBreakdown> countries;
    private final List<StatsBreakdown> referrers;
    private final List<StatsBreakdown> devices;
    private final List<StatsBreakdown> browsers;

    @SerializedName("top_codes")
    private final List<StatsBreakdown> topCodes;

    /**
     * Creates a {@code ProjectStats}.
     *
     * @param days        the window size in days the totals span
     * @param totalClicks total clicks recorded in the window
     * @param qrScans     total QR scans (a subset of {@code totalClicks})
     * @param daily       one point per day across the window
     * @param countries   top countries by click count
     * @param referrers   top referrers by click count
     * @param devices     click share by device class
     * @param browsers    click share by browser
     * @param topCodes    top short codes by click count
     */
    public ProjectStats(
            int days,
            long totalClicks,
            long qrScans,
            List<DailyClicksPoint> daily,
            List<StatsBreakdown> countries,
            List<StatsBreakdown> referrers,
            List<StatsBreakdown> devices,
            List<StatsBreakdown> browsers,
            List<StatsBreakdown> topCodes) {
        this.days = days;
        this.totalClicks = totalClicks;
        this.qrScans = qrScans;
        this.daily = daily;
        this.countries = countries;
        this.referrers = referrers;
        this.devices = devices;
        this.browsers = browsers;
        this.topCodes = topCodes;
    }

    private static <T> List<T> orEmpty(List<T> list) {
        return list == null ? Collections.emptyList() : list;
    }

    /** {@return the window size in days the totals span} */
    public int getDays() {
        return days;
    }

    /** {@return total clicks recorded in the window} */
    public long getTotalClicks() {
        return totalClicks;
    }

    /** {@return total QR scans (a subset of {@link #getTotalClicks()})} */
    public long getQrScans() {
        return qrScans;
    }

    /** {@return one point per day across the window; never {@code null}} */
    public List<DailyClicksPoint> getDaily() {
        return orEmpty(daily);
    }

    /** {@return top countries by click count; never {@code null}} */
    public List<StatsBreakdown> getCountries() {
        return orEmpty(countries);
    }

    /** {@return top referrers by click count; never {@code null}} */
    public List<StatsBreakdown> getReferrers() {
        return orEmpty(referrers);
    }

    /** {@return click share by device class; never {@code null}} */
    public List<StatsBreakdown> getDevices() {
        return orEmpty(devices);
    }

    /** {@return click share by browser; never {@code null}} */
    public List<StatsBreakdown> getBrowsers() {
        return orEmpty(browsers);
    }

    /** {@return top short codes by click count; never {@code null}} */
    public List<StatsBreakdown> getTopCodes() {
        return orEmpty(topCodes);
    }
}
